File: frontend/src/types/message.ts
export interface Message {
  id: string;
  channel_id: string;
  user_id: string;
  content?: string;
  message_type: 'text' | 'file' | 'system';
  reply_to?: string;
  thread_ts?: string;
  reply_count: number;
  is_edited: boolean;
  edited_at?: string;
  attachments?: any[];
  mentions?: string[];
  
  // User information
  user_username?: string;
  user_display_name?: string;
  user_avatar_url?: string;
  
  // Reaction information
  reaction_counts?: Record<string, number>;
  user_reactions?: string[];
  
  created_at: string;
  updated_at: string;
}

export interface MessageCreate {
  content: string;
  message_type?: 'text' | 'file' | 'system';
  reply_to?: string;
  attachments?: any[];
  mentions?: string[];
}

export interface MessageUpdate {
  content: string;
}

export interface MessageList {
  messages: Message[];
  total: number;
  has_more?: boolean;
}

export interface MessageReaction {
  id: string;
  message_id: string;
  user_id: string;
  emoji: string;
  created_at: string;
  
  // User information
  user_username: string;
  user_display_name: string;
  user_avatar_url?: string;
}

export interface MessageReactionCreate {
  emoji: string;
}

export interface ThreadResponse {
  parent_message: Message;
  replies: Message[];
  total_replies: number;
}

File: backend/app/models/message.py
"""
Message and related models.
"""
from sqlalchemy import Column, DateTime, ForeignKey, String, Text, Boolean, Integer, JSON
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from app.models.base import BaseModel


class Message(BaseModel):
    """Message model for chat messages."""
    
    __tablename__ = "messages"
    
    channel_id = Column(
        UUID(as_uuid=True),
        ForeignKey("channels.id"),
        nullable=False,
        comment="Channel where message was sent"
    )
    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=False,
        comment="User who sent the message"
    )
    content = Column(
        Text,
        nullable=True,
        comment="Message content (can be null for file-only messages)"
    )
    message_type = Column(
        String(20),
        nullable=False,
        default="text",
        comment="Message type: text, file, system"
    )
    reply_to = Column(
        UUID(as_uuid=True),
        ForeignKey("messages.id"),
        nullable=True,
        comment="Message this is a reply to"
    )
    thread_ts = Column(
        DateTime(timezone=True),
        nullable=True,
        comment="Thread timestamp for threaded messages"
    )
    reply_count = Column(
        Integer,
        default=0,
        comment="Number of replies in thread"
    )
    is_edited = Column(
        Boolean,
        default=False,
        nullable=False,
        comment="Whether message has been edited"
    )
    edited_at = Column(
        DateTime(timezone=True),
        nullable=True,
        comment="When message was last edited"
    )
    attachments = Column(
        JSON,
        nullable=True,
        comment="File attachments metadata"
    )
    mentions = Column(
        JSON,
        nullable=True,
        comment="List of mentioned user IDs"
    )
    
    # Relationships
    channel = relationship("Channel", back_populates="messages")
    # user = relationship("User", back_populates="messages")
    # parent_message = relationship("Message", remote_side="Message.id")
    # replies = relationship("Message", remote_side="Message.reply_to")
    # reactions = relationship("MessageReaction", back_populates="message")
    
    def __repr__(self) -> str:
        """String representation of Message."""
        content_preview = (self.content[:50] + "...") if self.content and len(self.content) > 50 else self.content
        return f"<Message(id={self.id}, user_id={self.user_id}, content='{content_preview}')>"
    
    def is_thread_reply(self) -> bool:
        """Check if this message is a reply in a thread."""
        return self.reply_to is not None


class MessageReaction(BaseModel):
    """Message reaction model."""
    
    __tablename__ = "message_reactions"
    
    message_id = Column(
        UUID(as_uuid=True),
        ForeignKey("messages.id"),
        nullable=False,
        comment="Message being reacted to"
    )
    user_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=False,
        comment="User who added the reaction"
    )
    emoji = Column(
        String(10),
        nullable=False,
        comment="Reaction emoji"
    )
    
    # Relationships
    # message = relationship("Message", back_populates="reactions")
    # user = relationship("User", back_populates="message_reactions")
    
    def __repr__(self) -> str:
        return f"<MessageReaction(id={self.id}, message_id={self.message_id}, emoji={self.emoji})>"